### ISSUE TITLE
Spectral initialization

### ISSUE THREAD

**Issue by dkobak** [pavlin-policar/openTSNE#110]
I think it'd be really cool to add `initialization='spectral'` support. It should be very easy, because it just means taking the P matrix after it's computed and running `scipy.sparse.linalg.svds` on it.

**Comment by pavlin-policar** [comment]
This is a really good idea. Is this what UMAP uses? I am a bit rusty, but I thought spectral embeddings were the top two eigenvectors of the graph Laplacian (A.T * A - D, where A is the adjacency matrix and D is the diagonal degree matrix), or is this some fancier spectral initialization?

I am totally for adding this though. I finally returned home, so I'll try to get a release out within the next few days.

**Comment by dkobak** [comment]
I'm not entirely sure what exactly UMAP is doing. I think we can do any sensible thing here, it does not have to be 100% equivalent to UMAP's choices.

There are several variants of spectral embeddings and several equivalent ways in which many can be defined. See this https://jlmelville.github.io/smallvis/spectral.html for a useful concise overview. If W is a symmetric (weighted) affinity matrix and D are its row/column sums, then one way to define Laplacian eigenmaps is to take the 2nd/3rd largest eigenvectors of D^{-1}W. 

In t-SNE's case, the directional affinities are row-normalized by construction but are not symmetric (so it's a Markov matrix). The final affinities P are symmetric but not row-normalized... So I guess one could take the largest eigenvectors of D^{-1}P, and not of P itself (as I wrote initially). Still very easy to implement.

Actually I would multiply the eigenvectors by their eigenvalues, as in diffusion maps: https://jlmelville.github.io/smallvis/spectral.html#diffusion_maps. Otherwise if you embed a line, then it will look like a parabola, even though the second eigenvalue is ~0. That's what https://scikit-learn.org/stable/modules/generated/sklearn.manifold.SpectralEmbedding.html does on a 1D line (because it does not multiply by the eigenvalues).

I think good checks would be to try it on a line and on a circle. It should recover the line and the circle.

**Comment by jlmelville** [comment]
UMAP uses the "symmetrized" normalized Laplacian, i.e. the one that isn't the Laplacian Eigenmaps graph Laplacian. I have never noticed any consistent advantage of one over the other, so my advice would be to choose the one that tends to lead to the most successful and quickest convergence over a range of datasets, which could be dependent on the solver used, e.g. there might be an advantage to using a symmetric matrix for some routines.

Also, have a fallback plan if the spectral routine fails, e.g. PCA. You probably do want the default convergence settings to be quite lax, but again this depends on the solver you use. I use RSpectra for uwot and it can unpredictably take an exceptionally long time to finish during which time the user can't cancel, which is a horrible user experience. Hopefully the Python routines are better for this. If not, it's probably better for it to finish quickly by default even if it fails to converge. Convergence issues can also be caused if the nearest neighbor graph isn't a single component. This isn't a big risk with default t-SNE settings, but I think I have seen it happen.

**Comment by dkobak** [comment]
Thanks for joining this discussion, James! These are good tips. That said, my suggestion was to keep using PCA as the default initialization, at least for the time being, so it won't be as critical as in UMAP if the spectral init takes a long time to converge or is somehow unstable.

**Comment by pavlin-policar** [comment]
All great tips, thanks! Will definitely keep them in mind.

**Comment by dkobak** [comment]
Are you planning to include this into the upcoming release? It would be great, I think. Let me know if you want a PR or prefer to implement it yourself.

**Comment by pavlin-policar** [comment]
Yeah, I think it would be a cool addition. I've been playing around with this for a little bit now, but I can't seem to get the line to work properly... I would appreciate it if you took a crack at it.

```python
import numpy as np
from sklearn.utils.extmath import svd_flip
from openTSNE.affinity import PerplexityBasedNN
import scipy.sparse as sp
import matplotlib.pyplot as plt

# Generate a line
n = 10000; X = np.zeros((n, 3)); X[:, 0] = np.arange(n)

affinities = PerplexityBasedNN(X, verbose=True)
W = affinities.P

D = np.sum(W, axis=1)
D = sp.diags(np.ravel(D), offsets=0)

U, S, VT = sp.linalg.svds((D.power(-1) @ W), k=3, solver="arpack")
S = S[::-1]
U, VT = svd_flip(U[:, ::-1], VT[::-1])

fig, ax = plt.subplots(ncols=2, figsize=(9, 4))
ax[0].scatter(U[:, 0], U[:, 1], s=1, c=np.arange(n))
ax[1].scatter(U[:, 1], U[:, 2], s=1, c=np.arange(n))
plt.show()
```

Plotting the first and second (left) and second and third eigenvectors (right)...
![image](https://user-images.githubusercontent.com/5758119/76681049-c3cf8380-65ee-11ea-977d-11e99ffe3757.png)

neither of which seems to be correct. The circle, however, appears to work, but only if I use the second and third eigenvectors, and not the first and second, as is written in the post.
![image](https://user-images.githubusercontent.com/5758119/76681068-15780e00-65ef-11ea-96c8-8dfd39a611a8.png)

I did also try using the lobpcg solver and TruncatedSVD from sklearn, which are much faster, but produce meaningless results, so go figure.

**Comment by dkobak** [comment]
The D^{-1}W matrix should have constant row sums, so the first eigenvector should be trivial and is supposed to be ignored. For the circle it works exactly as it should.

Not sure right now what happens with the line. The 1st eigenvector does not look trivial, which is strange.

**Comment by dkobak** [comment]
Not sure if it matters but I would put 1/D inside the diags.

**Comment by dkobak** [comment]
> The 1st eigenvector does not look trivial, which is strange.

I think it's a convergence issue. I tried 
```
S,U = sp.linalg.eigsh(W, M=D, k=3)
```
and it works fine in this example. This might be numerically more stable (?). 

Also, I must have been wrong that in this case the 3rd eigenvalue should be very small (because the manifold is 1-dimensional); it seems it's almost the same as the 2nd one. Then there is nothing gained by multiplying the eigenvectors with the eigenvalues. I am a bit surprised by this though. @jlmelville What do you think?

**Comment by jlmelville** [comment]
@dkobak, I'm afraid I don't have any intuition about the ratio of the eigenvalues in the 1D line case (or in any case for that matter).

In general, I don't know if this is an issue, but you have to be a bit careful about the ordering of eigenvectors if you use P instead of L_rw because you need to use one minus the eigenvalue before sorting them. Or ask for the largest instead of smallest eigenvectors and then ignore the top rather than the bottom eigenvector. It's confusing.

I always have to play around with some simple toy examples to persuade myself I know what's going on. In that spirit, if it's helpful, I have added a Python version of the R code I used for exploring the relationship between the graph Laplacians (for dense matrices only for now): <https://gist.github.com/jlmelville/8b7655fb4803ce49e4f560d316b04a46>

If you use a smaller `n`, you can combine that gist with your current experiments to play around a bit more without worrying about convergence issues and ordering and ensure that the correct eigenvector is the trivial one, e.g.:

```python
import matplotlib.pyplot as plt
import scipy.sparse as sp

from openTSNE.affinity import PerplexityBasedNN

n = 100; X = np.zeros((n, 3)); X[:, 0] = np.arange(n)

affinities = PerplexityBasedNN(X)
W = affinities.P
dW = np.array(sp.csr_matrix.todense(W))
WD = AffDeg(dW, degmat(dW))
lap = lapm(WD)

# effectively laplacian eigenmaps
le = eig(lap.Lrw)

fig, ax = plt.subplots(ncols=2, figsize=(9, 4))
ax[0].scatter(le.vectors[:, 1], le.vectors[:, 2], s=1, c=np.arange(n))
ax[1].scatter(le.vectors[:, 1] * le.values[1], le.vectors[:, 2] * le.values[2], s=1, c=np.arange(n))
plt.show()
```

Anyway, some of this might help with ensuring that the sparse version is returning the eigenvectors you want and it's not a convergence issue.
